packages/core/src/controller/resource-controller.ts

```diff
@@ -6,21 +6,23 @@ import { createFuture } from '../future';
 
 export function createResourceController<TOut>(task: Task<TOut>, resource: Resource<TOut>): Controller<TOut> {
   let delegate: Controller<TOut>;
-  let { resourceScope } = task.options;
+  let { scope } = task.options;
   let { produce, future } = createFuture<TOut>();
 
   function start() {
-    if(!resourceScope) {
+    if(!scope) {
       throw new Error('cannot spawn resource in task which has no resource scope');
     }
     let init;
     try {
-      init = resource.init(resourceScope, task);
+      init = resource.init(scope, task);
     } catch(error) {
       produce({ state: 'errored', error });
       return;
     }
-    delegate = createIteratorController(task, init, { resourceScope });
+    let name = resource.name || resource.labels?.name || 'resource';
+    let resourceTask = scope.run(undefined, { labels: { name, type: 'resource' } });
+    delegate = createIteratorController(task, init, { resourceTask });
```

Comment:
I know this is the way it was, but why must the `init()` function be a generator? Can we not just use `createController()` here?

Reply:
I think we probably could, though it does introduce somewhat of a circular dependency. I think the intention originally was to keep it simple, and only allow generators with an eye towards possibly generalizing this later on. That being said, I don't think there is a particularly strong use case for using something other than a generator.

Reply:
I agree that it will be rare to write an init function with something other than with a generator , but I think the strongest reason to make it more general is to reduce systemic complexity. That way, we can say "yes, `init()` is magical, but only in this very specific way"

Reply:
That's a good point. I'd rather tackle this in a separate PR though.